### ISSUE TITLE
ModelState cannot be added to TempData (cannot be serialized)

### ISSUE THREAD

**Issue by Bartmax** [aspnet/Mvc#4190]
I found this when dealing with `ModelState` and transfering using `TempData` to follow `PRG` pattern.

``` c#
TempData.Add("ModelStateTransfer", ModelState);


InvalidOperationException: 
The 'Microsoft.AspNet.Mvc.ViewFeatures.SessionStateTempDataProvider' cannot serialize an object
of type 'Microsoft.AspNet.Mvc.ModelBinding.ModelStateDictionary' to session state.
```

Shouldn't be `ModelState` serializable?

I was pointed out to this tests:
https://github.com/aspnet/Mvc/blob/25eb50120eceb62fd24ab5404210428fcdf0c400/test/Microsoft.AspNetCore.Mvc.Core.Test/SerializableErrorTests.cs

so I tried using `SerializableError` with no success

``` C#
TempData.Add("ModelStateTransfer", new SerializableError(ModelState));

InvalidOperationException: 
The 'Microsoft.AspNet.Mvc.ViewFeatures.SessionStateTempDataProvider' cannot serialize an object
of type 'Microsoft.AspNet.Mvc.SerializableError' to session state.
```

Help? Ideas?

**Comment by dougbu** [comment]
The `ModelStateDictionary` may contain `Exception`s and those are not necessarily serializable. For many scenarios, the `BadRequest(ModelStateDictionary)` method does the right thing. It substitutes a default error message and ignores the `Exception`s.

But your scenario might not be a good fit for returning the above information to the client. Please let us know more about your scenario if that's the case.

**Comment by Bartmax** [comment]
Basically, I want to save ModelState on a POST action so I can redirect and have the ModelState on next (get) request. Basic PRG pattern.

```
[HttpGet]
public IActionResult Create(){
    if (TempData.ContainsKey("ModelStateTransfer") {
        ModelState.Merge(TempData["ModelStateTransfer"] as ModelStateDictionary);
    }
    return View();
}

[HttpPost]
public IActionResult Create(CreateBidningModel binding) {
    if (!ModelState.IsValid) {
        TempData.Add("ModelStateTransfer", ModelState);
        return RedirectToAction(nameof(Create));
    }
    // ....
}
```

**Comment by Eilon** [comment]
In my apps when I've used PRG what I end up doing is that for *failed* requests it's still just a POST, but upon ultimate success (all fields correctly validated) I then do a redirect to a GET. The PRG pattern was intended to prevent re-submitting a form, which when the form data is invalid, is of course not an issue. And once the form data is valid, ModelState generally has nothing interesting in it (no errors, no exceptions, no validation messages, no distinct entered data, etc.).

I'm not familiar with the particular pattern used here. I actually don't think it works super well anyway because if there were ModelState errors and the user is redirected back to the initial GET page and then refresh their browser the TempData will be gone and they won't see any of the model state at all.

**Comment by danroth27** [comment]
No plans to change anything here.

**Comment by thiagomajesk** [comment]
@Eilon I don't think I quite understood your explanation. The reason to preserve ModelState here is to provide form usability. Imagine a huge form, if you treat failed requests just as a normal POST, then the user will have to fill every single form field again.

The way I see it, we have two options when `ModelState.IsValid = false`:
- Rebuilds the model and return view (duplicates code on GET and POST actions)
- (PRG) Redirects to GET that has the responsability to build the view

Of course the second options is preferable because of DRY. What is your approach right now? Anyone else could bring some thoughts on this matter?

**Comment by dougbu** [comment]
@thiagomajesk the usual choice is close to your first bullet. There's no need to "rebuild the model" because the view uses the (invalid) information from `ModelState`. And, and there's minimal duplication because the view can be reused.

E.g. from the MusicStore [sample](https://github.com/aspnet/MusicStore/blob/dev/samples/MusicStore/Controllers/ManageController.cs#L72-L88):
``` c#
        //
        // GET: /Account/AddPhoneNumber
        public IActionResult AddPhoneNumber()
        {
            return View();
        }

        //
        // POST: /Account/AddPhoneNumber
        [HttpPost]
        [ValidateAntiForgeryToken]
        public async Task<IActionResult> AddPhoneNumber(AddPhoneNumberViewModel model)
        {
            if (!ModelState.IsValid)
            {
                return View(model);
            }

            // ...
        }
```

**Comment by thiagomajesk** [comment]
@dougbu For simple types like `int` and `string` yes, I'm aware. 
But if I have a list of an complex type, when post happens, it will always be null (pretty obvious). In those cases we have to rebuild the view model to display the correct data again.

That's when **PRG** shines, because only the **GET** action needs to know how to build your views. 
If I'm always checking model state on **POST**, I'll have to duplicate the code that builds the view every single time (both o **GET** and **POST**). 

All that beeing said, what are your thoughts on reusing that code? I know that a view model builder or something like that is an options, but I'm wondering what other approaches could fit that case instead of **PRG**.

**Comment by Bartmax** [comment]
@dougbu post model is not always same as get model. This should also be encouraged more.

@thiagomajesk I feel like not supporting this scenario is not understanding the real issue but in done fighting it :) 

Maybe a blog post will help explain all get/post scenarios and why PRG is not about simple crud stuff. 

Also some views can have multiple forms to multiple actions with different models. This particular simple scenario is kinda wonky right now.

**Comment by thiagomajesk** [comment]
@Bartmax That's true. **PRG** by far seems the best way of dealing with this kinda scenarios.

Recently I came across this [article](http://andrewlock.net/post-redirect-get-using-tempdata-in-asp-net-core/) that demonstrates how to push the new `ModelState` into `TempData` (as we used to do in previous versions of asp.net).

One thing that got me thinking is that a guy in the comment section alleges that this functionality is supported on aspnet core 1.1 through cookie tempdata provider. Would love someone providing any thoughts on the matter.

**Comment by Bartmax** [comment]
That's similar of what I ended up doing. But that isn't a supported scenario on 1.1. That's just read the data serialize and create the model again in a very manual fashion. While it works, I still think it should be baked in the framework and not responsibility for the developer or external Nuget.

**Comment by ctolkien** [comment]
I'm going to chime in here as well with a 👍  - this is one of those death by a thousand cuts that makes MVC painful to use. IMO, it's fine when you can post to the same URL that you're currently on, it falls apart when you need to start posting to different URL's and handling error conditions.

**Comment by rynowak** [comment]
I'm reactivating this for consideration. We might not do exactly **this** but we should do something for this scenario

**Comment by Eilon** [comment]
I still don't believe that TempData is a solution for this. One hit of <kbd>F5</kbd> and your TempData goes away.

**Comment by Bartmax** [comment]
> I still don't believe that TempData is a solution for this. One hit of F5 and your TempData goes away.

🤦‍♂️

**Comment by thiagomajesk** [comment]
@Eilon Same thing for every other form... If the user refreshes the page, it **is** expected that everything goes away (errors, messages, values he's typed). Normally browsers will warn you about unsaved form changes. I think TempData is the perfect place for it, exactly because of what you just said. TempData's lifecyle is short and we don't want to keep those values hanging on the request any longer than that.

**Comment by Eilon** [comment]
The problem I describe doesn't happen with the Post-Redirect-Get (PRG) pattern because error cases are still a POST, so if you refresh, first of all the browser warns you, and even if you do post, you just re-post the same (bad) data, in which case you see the same errors again. If you do an initial redirect w/ TempData then the browser might not warn you, and if you proceed, you lose your data.

Is the problem I'm describing not a concern? We can certainly implement what's described here, it just doesn't seem to me like it would promote a good UI for end-users.

**Comment by Eilon** [comment]
We'll consider opening up the TempData default serializer to support whatever Json.NET supports, though there are no plans to do that at this time.

We can also consider making the TempData serializer itself pluggable (which I think it was in MVC 5.x).

**Comment by cgountanis** [comment]
Since this is the source thread I figured I would post would help me and it works great on any kind of model you create.

https://stackoverflow.com/questions/34638823/store-complex-object-in-tempdata-in-mvc-6

**Comment by Bartmax** [comment]
Side effects of this and not really understanding the value of proper PRG : https://github.com/aspnet/Announcements/issues/285

**Comment by ryanelian** [comment]
Laravel (PHP) has redirect function in controller action which can be fluently chained to include errors and form inputs.

https://laravel.com/docs/5.1/validation#other-validation-approaches

```php
class ProductController extends Controller {
    public function post(Request $request) {
        // make $validator here...

        if ($validator->fails()) {
            return redirect('post/create')
                    ->withErrors($validator)
                    ->withInput();
        }

        // else: do something...
    }
}
```

Related to issue topic, it'd be nice if we can do this, instead of manually playing with `TempData` / manually serializing `ModelState`

```cs
return Redirect("~/tosomewhere").WithModelState();
```

**Comment by thiagomajesk** [comment]
@ryanelian Love the idea! Definitely a lot cleaner than using attributes 👍.

**Comment by valeriob** [comment]
It's astonishing that MVC up to this day does not have a Post Redirect Get story...

**Comment by Bartmax** [comment]
The recommendations to expose over-posting vulnerability by using the full `ViewModel` on post and return a `View` on this thread is alarming.

**Comment by cccuscott** [comment]
Plus one for needing this functionality.
